File: Chapter-11/westbrook_shots_sim.py
###########################################################################
### Chapter 11 - Streakiness in Sports                                  ###
# Mathletics: How Gamblers, Managers, and Sports Enthusiasts              #
# Use Mathematics in Baseball, Basketball, and Football                   #
###########################################################################

import numpy as np
import pandas as pd
from tabulate import tabulate

game_shots = pd.read_csv("westbrook_shots.csv")

westbrook_fgp = sum(game_shots.FG)/sum(game_shots.FGA)

# simulate every game 
makes = list(np.zeros(10000))
shots = list(np.zeros(10000))
miss_after_three_makes = list(np.zeros(10000))
miss_after_three_misses = list(np.zeros(10000))
make_after_three_makes = list(np.zeros(10000))
make_after_three_misses = list(np.zeros(10000))

for b in range(10000):
	for i in range(len(game_shots)):
		s = [np.random.binomial(1,westbrook_fgp) for _ in range(game_shots['FGA'].iloc[i])]
		makes[b] += sum(s)
		shots[b] += len(s)
		for k in range(len(s)-3):
			if s[k:k+3] == [1,1,1]:
				try:
					if s[k+4] == 1:
						make_after_three_makes[b] += 1
					else:
						miss_after_three_makes[b] += 1
				except IndexError:
					# in this case the 4th shot of the sequence was not realized in the dataset. I.e., the simulation of the game finished with three makes, so we cannot use this observation.
					pass
			if s[k:k+3] == [0,0,0]:	
				try:
					if s[k+4] == 1:
						make_after_three_misses[b] += 1
					else:
						miss_after_three_misses[b] += 1
				except IndexError:
					# in this case the 4th shot of the sequence was not realized in the dataset. I.e., the simulation of the game finished with three misses, so we cannot use this observation. 
					pass


print(tabulate([[np.mean(makes),np.mean(shots)-np.mean(makes),np.mean(makes)/np.mean(shots),np.mean(make_after_three_makes),np.mean(miss_after_three_makes),np.mean(make_after_three_makes)/(np.mean(make_after_three_makes)+np.mean(miss_after_three_makes))]], headers=['Shots Made','Shots Misses','actual percentage','Made after 3 makes','Misses after 3 makes','percentage after 3 makes'], tablefmt='orgtbl'))
print(tabulate([[np.mean(makes),np.mean(shots)-np.mean(makes),np.mean(makes)/np.mean(shots),np.mean(make_after_three_misses),np.mean(miss_after_three_misses),np.mean(make_after_three_misses)/(np.mean(make_after_three_misses)+np.mean(miss_after_three_misses))]], headers=['Shots Made','Shots Misses','actual percentage','Made after 3 misses','Misses after 3 misses','percentage after 3 misses'], tablefmt='orgtbl'))

File: Chapter-11/westbrook_shots_sim.R
###########################################################################
### Chapter 11 - Streakiness in Sports                                  ###
# Mathletics: How Gamblers, Managers, and Sports Enthusiasts              #
# Use Mathematics in Baseball, Basketball, and Football                   #
###########################################################################

library(readr)
library(dplyr)
library(knitr)

game_shots <- read_csv("westbrook_shots.csv")

westbrook_fgp <- sum(game_shots$FG) / sum(game_shots$FGA)

makes <- rep(0, 10000)
shots <- rep(0, 10000)
miss_after_three_makes <- rep(0, 10000)
miss_after_three_misses <- rep(0, 10000)
make_after_three_makes <- rep(0, 10000)
make_after_three_misses <- rep(0, 10000)

for (b in 1:10000) {
  for (i in 1:nrow(game_shots)) {
    s <- rbinom(game_shots$FGA[i], 1, westbrook_fgp)
    makes[b] <- makes[b] + sum(s)
    shots[b] <- shots[b] + length(s)
    
    for (k in 1:(length(s) - 3)) {
      if (identical(s[k:(k+2)], c(1, 1, 1))) {
        if (!is.na(s[k + 4])) {
          if (s[k + 4] == 1) {
            make_after_three_makes[b] <- make_after_three_makes[b] + 1
          } else {
            miss_after_three_makes[b] <- miss_after_three_makes[b] + 1
          }
        }
      }
      
      if (identical(s[k:(k+2)], c(0, 0, 0))) {
        if (!is.na(s[k + 4])) {
          if (s[k + 4] == 1) {
            make_after_three_misses[b] <- make_after_three_misses[b] + 1
          } else {
            miss_after_three_misses[b] <- miss_after_three_misses[b] + 1
          }
        }
      }
    }
  }
}

results_after_three_makes <- data.frame(
  'Shots Made' = mean(makes),
  'Shots Misses' = mean(shots) - mean(makes),
  'actual percentage' = mean(makes) / mean(shots),
  'Made after 3 makes' = mean(make_after_three_makes),
  'Misses after 3 makes' = mean(miss_after_three_makes),
  'percentage after 3 makes' = mean(make_after_three_makes) / (mean(make_after_three_makes) + mean(miss_after_three_makes))
)

results_after_three_misses <- data.frame(
  'Shots Made' = mean(makes),
  'Shots Misses' = mean(shots) - mean(makes),
  'actual percentage' = mean(makes) / mean(shots),
  'Made after 3 misses' = mean(make_after_three_misses),
  'Misses after 3 misses' = mean(miss_after_three_misses),
  'percentage after 3 misses' = mean(make_after_three_misses) / (mean(make_after_three_misses) + mean(miss_after_three_misses))
)

print(results_after_three_makes)
print(results_after_three_misses)

